lte/gateway/c/session_manager/LocalEnforcer.cpp

```diff
@@ -247,58 +247,59 @@ void LocalEnforcer::sync_sessions_on_restart(std::time_t current_time) {
+  // for easy access
+  std::unordered_set<ImsiAndSid> sessions_with_active_flows;
   for (const RuleRecord& record : records.records()) {
-    auto it = session_map.find(record.sid());
-    if (it == session_map.end()) {
-      MLOG(MERROR) << "Could not find session for " << record.sid()
+    const std::string &imsi = record.sid(), &ip = record.ue_ipv4();
+    SessionSearchCriteria criteria(imsi, IMSI_AND_UE_IPV4, ip);
+    auto session_it = session_store_.find_session(session_map, criteria);
+    if (!session_it) {
+      MLOG(MERROR) << "Could not find session for " << imsi << " and " << ip
                    << " during record aggregation";
       continue;
     }
-    sessions_with_active_flows.insert(record.sid());
+    auto& session          = **session_it;
+    const auto& session_id = session->get_session_id();
     if (record.bytes_tx() > 0 || record.bytes_rx() > 0) {
-      MLOG(MINFO) << record.sid() << " used " << record.bytes_tx()
+      MLOG(MINFO) << session_id << " used " << record.bytes_tx()
                   << " tx bytes and " << record.bytes_rx()
                   << " rx bytes for rule " << record.rule_id();
     }
-    // Update sessions
-    for (const auto& session : it->second) {
-      SessionStateUpdateCriteria& uc =
-          session_update[record.sid()][session->get_session_id()];
-      session->add_rule_usage(
-          record.rule_id(), record.bytes_tx(), record.bytes_rx(), uc);
-    }
+    SessionStateUpdateCriteria& uc = session_update[imsi][session_id];
+    session->add_rule_usage(
+        record.rule_id(), record.bytes_tx(), record.bytes_rx(), uc);
+    sessions_with_active_flows.insert(ImsiAndSid(imsi, session_id));
   }
   complete_termination_for_released_sessions(
       session_map, sessions_with_active_flows, session_update);
 }
 
 void LocalEnforcer::complete_termination_for_released_sessions(
-    SessionMap& session_map, std::unordered_set<std::string> sessions_with_active_flows,
+    SessionMap& session_map,
+    std::unordered_set<ImsiAndSid> sessions_with_active_flows,
     SessionUpdate& session_update) {
   // Iterate through sessions and notify that report has finished. Terminate any
   // sessions that can be terminated.
-  std::vector<std::pair<std::string, std::string>> imsi_to_terminate;
+  std::vector<ImsiAndSid> imsi_to_terminate;
   for (const auto& session_pair : session_map) {
     const std::string imsi = session_pair.first;
     for (const auto& session : session_pair.second) {
       const std::string session_id = session->get_session_id();
       // If we did not receive a rule record for the session, this means
       // PipelineD has reported all usage for the session
       if (session->get_state() == SESSION_RELEASED &&
-          sessions_with_active_flows.find(imsi) == sessions_with_active_flows.end()) {
+          sessions_with_active_flows.find(ImsiAndSid(imsi, session_id)) ==
+              sessions_with_active_flows.end()) {
```

Comment:
I guess all this PR is around differentiating case of having `SESSION_RELEASED` but with active flows. But when does that happen? Can we have a session marked as released (which pipelined should have already removed the flows) but with active flows?

Reply:
So Pipelined maintains two sets of flows. The actual flows that act on the policy and enforcement stats flows that collect usage on those rules. When we go from ACTIVE->RELEASED, we remove the first set of rules. But PipelineD will keep reporting enforcement stats as long as they exist. Once the last bit of usage has been reported, pipelined removes the enforcement stats flows. 
So after we move from ACTIVE->RELEASED, we have to wait at least one reporting period up to two reporting periods to get the last chunk of usage.

Reply:
Maybe this logic deserves a diagram in the README :)

Reply:
yep, this is the issue we always have had with the last chunk of data.... 
Thanks!